### PR TITLE
Allow redefining None to 0 for @[Flags] enum

### PR DESCRIPTION
**Pull request by bew**
Followup on https://github.com/crystal-lang/crystal/pull/4395#issuecomment-301329038

#4395 now prevents `All` & `None` in `@[Flags]` enums.

But as @oprypin [pointed out / demonstrated](https://github.com/crystal-lang/crystal/pull/4395#issuecomment-301329038), redefining `None` & `All` can be useful for **documentation purpose**.

To this end:
> it should be possible to redefine `None` and `All` but only if their value matches what they would have been anyway.

`None` should always be `0`, and All should always be all the values OR-ed together.

So something like this should work:
```cr
@[Flags]
enum Foo
  None = 0
  A
  B
  C
  All = A | B | C
end
```
It's easy to handle `None`, just need to check it's `0`.
About `All`, first it's very tedious to repeat all members, second it's harder to detect if there are more members after it. To simplify `All` and still be able to documente it, there could be a special value like `All = :all` that will skip this member but retain its doc somehow...

---
As handling `None` is easy I implemented it in this PR.
Also, I don't think `All` is common when using enum flags, so I think it's ok to not be able to document it.

WDYT?

### CODE REVIEW

**Comment by oprypin** [comment]
Mentioning `:all` was a bad call because it's not even part of this PR 🐱.

I would be very happy to have just `None` merged.

**Comment by bew** [comment]
I just added a better empty enum check that ignores None & All

File: src/compiler/crystal/semantic/top_level_visitor.cr [hunk 1]
```diff
@@ -569,7 +569,13 @@ class Crystal::TopLevelVisitor < Crystal::SemanticVisitor
         is_flags: enum_type.flags?)
     end
 
-    if enum_type.types.empty?
+    nb_members = enum_type.types.size
```
[the remark this answers is not in the document]
Comment (bew) [reply src/compiler/crystal/semantic/top_level_visitor.cr]: it's the number of members, maybe `number_of_members` then? it just felt too long ^^

[the remark this answers is not in the document]
Comment (straight-shoota) [reply src/compiler/crystal/semantic/top_level_visitor.cr]: There is no "too long" for names, only "too short" 😄 `num_members` would be fine, tough. Or `members_size`.

[the remark this answers is not in the document]
Comment (bew) [reply src/compiler/crystal/semantic/top_level_visitor.cr]: `num_members` is good :smiley: ty

**Comment by oprypin** [comment]
There is no change in behavior of `All`.

**Comment by bew** [comment]
Yeah sorry if it wasn't clear @bcardiff, the title only mention the possibility to redefine None, not All. The PR inital msg is providing context and explanations, maybe it should have gone in an issue.

**Comment by bew** [comment]
Thank you!

**Comment by straight-shoota** [comment]
No, thank you, @bew =)

**Pull request merged**